lib/Corres_UL.thy

```diff
@@ -481,6 +481,18 @@ lemma corres_if3:
                                     (if G then a else b) (if G' then c else d)"
   by simp
 
+lemma corres_if_strong:
+  "\<lbrakk>\<And>s s'. \<lbrakk>(s, s') \<in> sr; R s; R' s'\<rbrakk> \<Longrightarrow> G = G';
+    \<lbrakk>G; G'\<rbrakk> \<Longrightarrow> corres_underlying sr nf nf' r P P' a c;
+    \<lbrakk>\<not> G; \<not> G'\<rbrakk> \<Longrightarrow> corres_underlying sr nf nf' r Q Q' b d \<rbrakk>
+   \<Longrightarrow> corres_underlying sr nf nf' r
+         (R and (if G then P else Q)) (R' and (if G' then P' else Q'))
+         (if G then a else b) (if G' then c else d)"
```

Comment:
It feels like we have loads of variations of `if` rules already, but I don't think we have anything where both guard conditions are in the assumptions. I was initially going to suggest deriving some of the other variations from this rule, but the proofs are all one-liners anyway, so I'm fine with it as it is.

Reply:
I think this is the only rule for `if` that also allows us to use the assumptions here to show equality of `G` and `G'`

Reply:
Oh sorry, I misread your comment. We're saying the same thing, oops

Reply:
All good, I should probably not even have said anything in the end, but I was too lazy to delete it again :-)